nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -391,9 +414,24 @@ in
               Type = "oneshot";
               RemainAfterExit = true;
             };
-            script = ''
-              zpool_cmd="${packages.zfsUser}/sbin/zpool"
-              ("$zpool_cmd" list "${pool}" >/dev/null) || "$zpool_cmd" import -d ${cfgZfs.devNodes} -N ${optionalString cfgZfs.forceImportAll "-f"} "${pool}"
+            script = (importLib {
+              zpoolCmd="${packages.zfsUser}/sbin/zpool";
+              awkCmd="${pkgs.gawk}/bin/awk";
+              inherit pool cfgZfs;
+            }) + ''
+              imported && exit
+              echo -n "importing ZFS pool \"${pool}\"..."
+              # Loop across the import until it succeeds, because the devices needed may not be discovered yet.
+              for trial in `seq 1 60`; do
+                ready && import && break
+                sleep 1
+              done
+              imported || import  # Try one last time, e.g. to import a degraded pool.
```

Comment:
It isn't tried in the same way. "ready && import" would only import the pool if *every* device is available, i.e. it's in state ONLINE and not state DEGRADED.

The unconditional import at the end will fail if it's in state FAULTED or MISSING, of course, but not if it's DEGRADED. That's a feature: The system will still boot if one or two devices are missing, just not as quickly. But we don't want it to import the pool in a degraded state if it can avoid that by waiting another few seconds.